src/overlays/actors/ovl_Boss_01/z_boss_01.c

```diff
@@ -246,191 +647,2927 @@ ActorInit Boss_01_InitVars = {
+        gSPDisplayList(POLY_XLU_DISP++, gOdolwaEyeDL);
+        Matrix_Pop();
+
+        CLOSE_DISPS(play->state.gfxCtx);
+    }
+}
+
+void Boss01_Draw(Actor* thisx, PlayState* play) {
+    static Vec3f sDefaultPelvisColliderOffset = { 10000.0f, 10000.0f, 10000.0f };
+    Boss01* this = THIS;
+    s32 pad;
+    u8* tex = GRAPH_ALLOC(play->state.gfxCtx, ODOLWA_SHADOW_TEX_SIZE);
+
+    OPEN_DISPS(play->state.gfxCtx);
+
+    // When Odolwa is performing a horizontal slash, an additional collider offset from his pelvis will sweep across the
+    // floor and damage the player. However, this collider should not be present all the time, so it is offset very far
+    // out-of-bounds here. If Odolwa is performing a horizontal slash, then Boss01_PostLimbDraw (which is called later
+    // in this function) will offset it properly for this frame.
+    Boss01_SetColliderSphere(ODOLWA_SWORD_COLLIDER_PELVIS, &this->swordCollider, &sDefaultPelvisColliderOffset);
+
+    Gfx_SetupDL25_Opa(play->state.gfxCtx);
+    Gfx_SetupDL25_Xlu(play->state.gfxCtx);
+
+    if (this->damageFlashTimer & 1) {
+        POLY_OPA_DISP = Gfx_SetFog(POLY_OPA_DISP, 255, 0, 0, 255, 900, 1099);
+    }
+
+    SkelAnime_DrawFlexOpa(play, this->skelAnime.skeleton, this->skelAnime.jointTable, this->skelAnime.dListCount,
+                          Boss01_OverrideLimbDraw, Boss01_PostLimbDraw, &this->actor);
+
+    POLY_OPA_DISP = Play_SetFog(play, POLY_OPA_DISP);
+
+    Boss01_GenShadowTex(tex, this, play);
+    Boss01_DrawShadowTex(tex, this, play);
+
+    if (sSwordTrailAlpha > 0.0f) {
+        Boss01_DrawSwordTrail(this, play);
+        Math_ApproachZeroF(&sSwordTrailAlpha, 1.0f, 50.0f);
+    }
+
+    Actor_DrawDamageEffects(play, &this->actor, this->bodyPartsPos, 15, this->drawDmgEffScale,
+                            this->drawDmgEffFrozenSteamScale, this->drawDmgEffAlpha, this->drawDmgEffType);
+
+    Boss01_DrawEffects(play);
+
+    CLOSE_DISPS(play->state.gfxCtx);
+}
+
+void Boss01_Afterimage_Draw(Actor* thisx, PlayState* play) {
+    Boss01* this = THIS;
+    s32 pad;
+    Boss01* parent = (Boss01*)this->actor.parent;
+
+    OPEN_DISPS(play->state.gfxCtx);
+
+    Gfx_SetupDL25_Opa(play->state.gfxCtx);
+    Gfx_SetupDL25_Xlu(play->state.gfxCtx);
+    POLY_OPA_DISP = Gfx_SetFog(POLY_OPA_DISP, 50, 0, 40, 255, 900, 1099);
+    SkelAnime_DrawFlexOpa(play, parent->skelAnime.skeleton, this->skelAnime.jointTable, parent->skelAnime.dListCount,
+                          NULL, NULL, &this->actor);
+    POLY_OPA_DISP = Play_SetFog(play, POLY_OPA_DISP);
+
+    CLOSE_DISPS(play->state.gfxCtx);
+}
+
+/**
+ * These four arrays encode circular shadow maps of various sizes. For an array of length N, the shadow map is N rows
+ * tall, and each entry in the array describes the start and end point of the shadow within a given row (the start point
+ * is just the negative end point). To illustrate using the sShadowSmallMap as an example:
```

Comment:
Not sure this is quite right. I don't think the endpoint actually gets included in the map (note the array ends at < endpoint, so would end one early). Also at least for Odolowa the last two maps end with `< endpoint - 1`.

Reply:
Huh, you're right. Wouldn't that mean the shadows are asymmetrical?

Reply:
Take a look at the most recent commit and let me know if you think that comment is correct. It is super strange to me that these shadows are asymmetrical, but it sure seems like they are.

Reply:
No I think they are still symmetrical just not about (`0`). For example the small map I think would look like the below, since the highest value is `3` and we have ` < 3 - 1` so the last index valid would just be `1`.
```
 * -3 -2 -1  0  1
 *  -----------------
 *  0  0  1  0  0
 *  0  1  1  1  0
 *  1  1  1  1  1
 *  1  1  1  1  1
 *  0  1  1  1  0
 *  0  0  1  0  0 
```

Reply:
Ahhhhhh, I reread the code again more closely, and yeah that makes sense. Thanks for giving it another look! Let me update the comment one last time.